### ISSUE TITLE
[EE5/6?] Issue in (simple) search: some custom field/s not searched when search keywords are quoted?

### ISSUE THREAD

**Issue by PaulBaileyAtChan** [ExpressionEngine/ExpressionEngine#1851]
(This is a tricky one to characterize and reproduce. Also, I'm describing what I see in EE5.4.3. I'm not in a position right now to test in EE6, but I don't see anything in the EE6 change logs which directly refers to the issue. So, please read this as something I'm seeing in EE5, which might also be happening in EE6, but needs to be confirmed.)

I'm writing some code to post-process (simple) search results, and I hit some behaviour in simple search results which looks wrong. Specifically, when search terms are quoted, some search hits which should be included are missing, and this might relate to a failure to search all custom fields. To help diagnose, I've exported some of the SQL queries generated by the search module.

A simple example. Searching for a specific last name, **mcguire**, results in 3 hits. One of these occurrences is in a (searchable) custom field. Repeating the search with a quoted search term, **"mcguire"**, results in 2 hits; the missing hit in this case is the one in the custom field. The two SQL search queries generated by these searches are attached. Note that the second (for the quoted search term) is significantly smaller than the first. Specifically, note that the relevant custom field, **326**, is present in the SQL query for the unquoted search term, and missing from the SQL query for the quoted search term.

This behaviour looks to be the same for more complex searches. For example, a search for a specific **firstname lastname** includes the custom field when the search term is unquoted, but doesn't include the custom field when the search term is quoted, as **"firstname lastname"**.

Exported SQL queries follow.

SQL query when searching for **mcguire**:
[search-queries-mcguire-noquotes.txt](https://github.com/ExpressionEngine/ExpressionEngine/files/8149819/search-queries-mcguire-noquotes.txt)

SQL query when searching for **"mcguire"**:
[search-queries-mcguire-quotes.txt](https://github.com/ExpressionEngine/ExpressionEngine/files/8149820/search-queries-mcguire-quotes.txt)

**Comment by JayGreentree** [comment]
This is a known bug and a fix should be available soon.

**Comment by intoeetive** [comment]
@PaulBaileyAtChan what field type is the referenced `326` field? Just noting that I wasn't able to replicate the issue using textarea field.

Also just making sure you're running both tests on the same form and it has `search_in="everywhere"` parameter set

**Comment by PaulBaileyAtChan** [comment]
Yes, the tests are using exactly the same form. For what it's worth, it uses `search_in="entries"` (there are no comments), but I just tried with `search_in="everywhere"` and the behaviour is the same.

Field 326 is a textarea. I've just checked, and the other searchable fields in the same field group are also missing from the search. Fields 323 (textarea) and 324 (text input) are also both searchable, and appear in the search query when the search term isn't quoted, but are missing when it is quoted.

So it looks like that entire field group is missing from the search when the search term is quoted. It looks like the channel itself is being searched (it's the only channel which uses that field group), because the channel ID appears in the search query, but none of the searchable fields for that channel appear.

I definitely get that this is a hard one to nail down!

**Comment by intoeetive** [comment]
Are these fields assigned to the channel as part of field group, or individually?

**Comment by PaulBaileyAtChan** [comment]
The fields are assigned as part of a field group.

I do think the key here is comparing the example SQL queries that I attached. Large sections are missing from the query for the quoted search that are present in the query for the unquoted search. The code which generates the missing parts of the query looks to run between lines 752 and 777 of mod.search.php. Note that the query fragment `CAST(CONCAT` (line 760) is used a lot in the SQL query for the unquoted search, but doesn't appear at all in the SQL for the quoted search. That looks significant.

**Comment by intoeetive** [comment]
I am testing this in EE6, so there's a chance this has been fixed without proper record in changelog...

Can you post the template code for the form that you use?

**Comment by PaulBaileyAtChan** [comment]
The template code is pretty standard:

```
{exp:search:simple_form search_in="entries" results="10" status="open" where="all" channel="list-of-channels" result_page="search/index" no_result_page="search/fail"}
<input type="text" name="keywords" id="keywords" maxlength="100" />
{/exp:search:simple_form}

```

But I really don't want to be wasting your time here. I'm not in a position to test with EE6 at the moment, but I should be in a few weeks. If things look to be fixed for me in EE6, I'll close this. If not, I'll try to dig a little deeper. Either way, it's probably best to leave it for now. But thanks for looking!

**Comment by PaulBaileyAtChan** [comment]
@intoeetive Okay, I have some more information. I was able to run EE6.3.1 and EE5.4.3 side by side, with the same databases, and the same queries.

In EE6.3.1, quoting a search query doesn't seem to make a difference to the results (at least with my search tag). This is a change from EE5. *But*, the results aren't correct. Either way (quoted or unquoted), the results I get in EE6 match those in EE5 when the search term is quoted — that is, some results are clearly missing.

So, I poked around in the code some more (mod.search.php), and landed on:

<img width="967" alt="Screen Shot 2022-04-13 at 3 46 38 PM" src="https://user-images.githubusercontent.com/46760375/163282263-bb7d83a1-6cc8-4c84-9d61-8990932960d4.png">

My reading of these lines is that only 50 custom fields are being referred to. Line 384 divides the custom fields into chunks of 50, and then line 388 only appears to take the first chunk of 50. I don't see an enclosing loop which would loop over each chunk.

Guessing that this causes some custom fields to not be included in the search, I bumped up the chunk size to 500 — which easily encompasses all of my custom fields:

<img width="870" alt="Screen Shot 2022-04-13 at 3 52 48 PM" src="https://user-images.githubusercontent.com/46760375/163282792-0772c17c-5e31-4fa0-bd19-b863ee233939.png">

Running with this larger chunk size, I get search results with nothing obviously missing — the results match the results I get in EE5 with *unquoted* search keywords.

So, there might be two separate issues here, I'm not sure. But what I'm finding is that *all* searches in EE6 (quoted or not) are failing to search some of my custom fields, and that increasing the chunk size appears to fix things.

This needs a more careful look, obviously — there's likely stuff I'm missing (and I have no idea why the array chunking is there at all). But this might be a place to start.

**Comment by PaulBaileyAtChan** [comment]
Without intending to nag, if just changing the custom field array chunk size (as shown above) changes the search results, as it seems to, this looks enough to at least flag the issue as a significant bug.

**Comment by studionigh** [comment]
I am having a very similar problem.  I have set the parameter: search_in="everywhere". however, it is only searching the entry titles. I noticed I did not have a custom field {sub_title} set to "Include in search?" so I activated that, and the simple search still doesn't search that field. I have also reindexed the search in the utilities.

**Comment by studionigh** [comment]
WOW! This was a fix for me too!  Changed the 50 to 500, and now the search works correctly!

**Comment by PaulBaileyAtChan** [comment]
Great. Glad it's not just me! Will try to push for a fix to this, because it definitely looks like a bug.

**Comment by bici** [comment]
i implemented this fix posted by Bailey on the latest  Version 7 of EE. I am surprised that this has not been addressed as it seems to be a relatively easy code fix. I say easy but there may be other underlying reasons why it has not been addressed. 
**NOTE**: only by implementing this fix have i been able to successfully search all entires in all channels. without the code fix searching was broken on my site.

**Comment by robinsowell** [comment]
Definitely seem to be a couple of problems.

First, not sure why we need ALL of the custom fields.  I was debugging one, search had a channel parameter on it, only searching 'blog', which had 1 group with 3 fields in it.  So it really only needs to include those 3.  But it was pulling more than 50 and then getting truncated.  Not sure there's a reason for that, but worth a look.

The reason it's truncated- you'll see it in the channel module too.  You can only join so many tables in mysql- and in modern ee, each field is a table.  You'll hit a mysql error if you have hundreds and include them all in a join.  BUT it also shouldn't just be taking a random 50 and calling it good.  It should batch them, like channel entries does.  Assuming it's using them in the same way, which... it probably is.

We'll definitely dig into it more, but I don't think we can safely bump that array cutoff without a deeper dive.  If it's working for a quick patch?  Then that's totally fine for short term.

**Comment by PaulBaileyAtChan** [comment]
@robinsowell Thanks for marking it as accepted. :-) I have no clue about the MySQL restrictions, but batching together results sounds like the right approach. Looking at the code, it's almost as if that was attempted as a solution at some point, but only partially implemented.

I'd be wary of seeing my patch as a reason to defer looking at this seriously. It does feel like a significant bug to me — not least because it can be pretty hard to characterize, or even notice what's happening. And most users won't know to come to Github to find an undocumented patch.

**Comment by intoeetive** [comment]
I investigated this and it looks like switching to approach used by channel:entries tag would require a lot of code changes. So I can't say where that lays in priorities for us, considering that we have now much more powerful Pro Search module bundled with ExpressionEngine

**Comment by silenz88** [comment]
I think this one needs another review.

At the point, at which the array_chunk() cuts off the 50 fields they have not been filtered for being searchable yet.

This happens directly after (where the JOINs are made).

So you might have 100 fields from which only 3 are searchable but end up with none being searched at all.

**Comment by intoeetive** [comment]
@silenz88 the fix for this will be released in 7.3.0 (we don't have release date set for that yet, as we aim to also include few other changes)
If you need a fix right now, you may try pre-release from https://github.com/ExpressionEngine/ExpressionEngine/releases 
(But please be aware that we don't currently support updating from pre-release to released; one would need to roll back the version and run update again once the release is out)
